### PR TITLE
Add forward compatibility tests in CI

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64139

This PR adds CI test that can detect forward compatibility breaking changes for aten operator schemas. Since we don't have a strong policy on what constitutes as FC breaking change, we will just warn users when they introduce FC incompatible changes. This will later be replaced with more strict testing once we implement operator versioning. 

Differential Revision: [D30626912](https://our.internmc.facebook.com/intern/diff/D30626912)

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

File: aten/src/ATen/core/function_schema_inl.h [hunk 1]
```diff
@@ -84,6 +84,29 @@ inline bool Argument::isBackwardCompatibleWith(
...
+  const Argument* lhs = this;
+  const Argument* rhs = &old;
+  if (!(lhs->name() == rhs->name() && lhs->N() == rhs->N() &&
+        lhs->alias_info() == rhs->alias_info())) {
+    return false;
+  }
+  if (lhs->kwarg_only() && !rhs->kwarg_only()) {
+    return false;
+  }
+  if (!rhs->type()->isSubtypeOfExt(lhs->type(), why_not)) {
```
Comment (gmagogsfm) [inline aten/src/ATen/core/function_schema_inl.h:100]: Should this be reversed?

IIUC, `rhs` stands for older version of schema, `lhs` stands for newer version of schema. When `lhs > rhs`, ie newer version of schema allows a type that is a super type of older version of schema, we would have forward compatibility problem. Let me know what you think.

File: aten/src/ATen/core/function_schema_inl.h [hunk 2]
```diff
@@ -169,6 +192,86 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
...
+inline bool FunctionSchema::isForwardCompatibleWith(
+    const FunctionSchema& old,
+    std::ostream* why_not) const {
+  if (!(name() == old.name() &&
+        overload_name() == old.overload_name()
+        // we are conservative on is_vararg and is_varret,
+        // since they are only used by internal operators
+        && is_vararg() == old.is_vararg() && is_varret() == old.is_varret() &&
+        returns().size() == old.returns().size() &&
+        arguments().size() >= old.arguments().size())) {
```
Comment (gmagogsfm) [inline aten/src/ATen/core/function_schema_inl.h:204]: The case of `arguments.size() < old.arguments().size()` can be FC actually. Think about the case where we remove the last argument, which has a default value.

What do you think?

File: aten/src/ATen/core/function_schema_inl.h [hunk 3]
```diff
@@ -169,6 +192,86 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
...
+  }
+
+  // find the start of out args in the old schema
+  int old_out_start_idx;
+  for (old_out_start_idx = old.arguments().size() - 1; old_out_start_idx > -1;
+       old_out_start_idx--) {
+    if (!old.arguments().at(old_out_start_idx).is_out()) {
+      break;
+    }
+  }
```
Comment (gmagogsfm) [inline aten/src/ATen/core/function_schema_inl.h:215]: Really need to make this into a helper func : )

File: aten/src/ATen/core/function_schema_inl.h [hunk 4]
```diff
@@ -169,6 +192,86 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
...
+      return false;
+    }
+
+    auto default_val = arguments().at(i).default_value().value();
+    if (default_val.isList() || default_val.isGenericDict()) {
+      if (why_not) {
+        *why_not
+            << "Function schema not forward compatible since the new argument '"
+            << arguments().at(i).name() << "' of type "
+            << arguments().at(i).type()->str() << " is a container type.";
```
Comment (gmagogsfm) [inline aten/src/ATen/core/function_schema_inl.h:257]: Add more information here about we don't support default values of container type.

File: test/test_function_schema.py [hunk 5]
```diff
@@ -112,6 +112,32 @@ def test_string_optional_parameter_default_value(self):
         schema_b = parse_schema(str(schema_a))
         self.assertEquals(schema_a, schema_b)
 
+    def test_forward_compatible_arguments_without_out(self):
+        old_schema = parse_schema('any(Tensor self, int a, int b=1) -> Tensor')
+        new_schema = parse_schema('any(Tensor self, int a, int b=1, int c=1) -> Tensor')
+        self.assertTrue(new_schema.is_forward_compatible_with(old_schema))
+        new_schema = parse_schema('any(Tensor self, int a, int b=1, int[2] c=1) -> Tensor')
+        self.assertFalse(new_schema.is_forward_compatible_with(old_schema))
```
Comment (gmagogsfm) [inline test/test_function_schema.py:120]: Looks like `isForwardCompatibleWith`'s PyBind definition doesn't allow returning a string of `why_not`. Could we add it and verify match failures are due to the right reasons?

**Review by gmagogsfm — CHANGES_REQUESTED** [review]

**Review by janeyx99 — COMMENTED** [review]
Looks fine from the CI side!

File: test/test_function_schema.py (hunk 5 above)
Comment (tugsbayasgalan) [reply test/test_function_schema.py:120]: That's a good idea. Will do

File: aten/src/ATen/core/function_schema_inl.h (hunk 4 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/core/function_schema_inl.h:257]: Sounds good.

File: aten/src/ATen/core/function_schema_inl.h (hunk 2 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/core/function_schema_inl.h:204]: That is true. I will update it

File: aten/src/ATen/core/function_schema_inl.h (hunk 1 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/core/function_schema_inl.h:100]: Yeah you are right. It should be reversed. Will update

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

File: aten/src/ATen/core/function_schema.h [hunk 6]
```diff
@@ -199,6 +208,28 @@ struct FunctionSchema {
       const FunctionSchema& old,
       std::ostream* why_not = nullptr) const;
 
+  // Checks whether this schema is forward compatible with the old one.
```
Comment (mruberry) [inline aten/src/ATen/core/function_schema.h:211]: Cool list and super clear. Might be helpful to actually state which case is allowed/supported, too, which is adding a new kwarg with a default value at the end of a signature without out= or just before the out= kwarg on signatures that have it

File: aten/src/ATen/core/function_schema_inl.h [hunk 7]
```diff
@@ -169,6 +191,86 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
...
+    }
+
+    auto default_val = arguments().at(i).default_value().value();
+    if (default_val.isList() || default_val.isGenericDict()) {
+      if (why_not) {
+        why_not
+            << "Function schema is not forward compatible since the new argument '"
+            << arguments().at(i).name() << "' of type "
+            << arguments().at(i).type()->str() << " has a container type "
+            << "as its' default value.";
```
Comment (mruberry) [inline aten/src/ATen/core/function_schema_inl.h:251]: "its'" -> "its"

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 8]
```diff
@@ -197,6 +199,49 @@ def check_bc(existing_schemas):
         )
     return is_bc
 
+def check_fc(existing_schemas):
+    new_schema_dict = load_schemas_to_dict()
+    is_fc = True
+    broken_ops = []
+    # TODO skip allow list for now.
```
Comment (mruberry) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:206]: Can support for the skip list be added to this PR, or are there design issues? Like are you thinking of having a separate BC and FC allow lists?

File: test/test_function_schema.py [hunk 9]
```diff
@@ -112,6 +112,53 @@ def test_string_optional_parameter_default_value(self):
         schema_b = parse_schema(str(schema_a))
         self.assertEquals(schema_a, schema_b)
 
+    def test_forward_compatible_arguments_without_out(self):
+        old_schema = parse_schema('any(Tensor self, int a, int b=1) -> Tensor')
```
Comment (mruberry) [inline test/test_function_schema.py:116]: Add comments for each case being tested before it's tested

File: test/test_function_schema.py [hunk 10]
```diff
@@ -112,6 +112,53 @@ def test_string_optional_parameter_default_value(self):
...
+        new_schema = parse_schema('any(Tensor self, int a) -> Tensor')
+        is_fc, _ = new_schema.check_forward_compatible_with(old_schema)
+        self.assertTrue(is_fc)
+        new_schema = parse_schema('any(Tensor self, int a, int b=1, int c=1) -> Tensor')
+        is_fc, _ = new_schema.check_forward_compatible_with(old_schema)
+        self.assertTrue(is_fc)
+        new_schema = parse_schema('any(Tensor self, int a, int b=1, int[2] c=1) -> Tensor')
+        is_fc, reason = new_schema.check_forward_compatible_with(old_schema)
+        self.assertFalse(is_fc)
+        self.assertEqual(reason, "Function schema is not forward compatible since the new argument"
```
Comment (mruberry) [inline test/test_function_schema.py:126]: Is there a good test for type narrowing/widening?

File: test/test_function_schema.py [hunk 11]
```diff
@@ -112,6 +112,53 @@ def test_string_optional_parameter_default_value(self):
         schema_b = parse_schema(str(schema_a))
         self.assertEquals(schema_a, schema_b)
 
+    def test_forward_compatible_arguments_without_out(self):
+        old_schema = parse_schema('any(Tensor self, int a, int b=1) -> Tensor')
+        new_schema = parse_schema('any(Tensor self, int a) -> Tensor')
+        is_fc, _ = new_schema.check_forward_compatible_with(old_schema)
+        self.assertTrue(is_fc)
+        new_schema = parse_schema('any(Tensor self, int a, int b=1, int c=1) -> Tensor')
```
Comment (mruberry) [inline test/test_function_schema.py:120]: What about adding a new kwarg without a default value?

File: test/test_function_schema.py [hunk 12]
```diff
@@ -112,6 +112,53 @@ def test_string_optional_parameter_default_value(self):
...
+                                 " \'c\' of type int[] has a container type as its\' default value.")
+        new_schema = parse_schema('any(Tensor self, int a, int b=4) -> Tensor')
+        is_fc, reason = new_schema.check_forward_compatible_with(old_schema)
+        self.assertFalse(is_fc)
+        self.assertEqual(reason, "\'b\' is not forward compatible with the older version of the schema")
+        new_schema = parse_schema('any(Tensor self, int a, int c=1) -> Tensor')
+        is_fc, reason = new_schema.check_forward_compatible_with(old_schema)
+        self.assertFalse(is_fc)
+        self.assertEqual(reason, "\'c\' is not forward compatible with the older version of the schema")
+        new_schema = parse_schema('any(Tensor self, int a, int c=1, int b=1) -> Tensor')
```
Comment (mruberry) [inline test/test_function_schema.py:136]: What about a positional arg becoming kwarg-only and a kwarg-only arg becoming positional?

**Review by mruberry — COMMENTED** [review]
Overall looks pretty good! I left some comments inline for your review, @tugsbayasgalan. Looking forward to hearing your thoughts!

File: aten/src/ATen/core/function_schema_inl.h (hunk 7 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/core/function_schema_inl.h:251]: hhha will do :)

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 8 above)
Comment (tugsbayasgalan) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py:206]: The skip list was for BC before, so I was hesitant to include them here. But now I think about it, it probably makes more sense for those schemas to be considered for both back and forward compatibility. I will allow skip list here then. I think we should maintain one single list for both BC and FC. @gmagogsfm What do you think?

File: test/test_function_schema.py (hunk 10 above)
Comment (tugsbayasgalan) [reply test/test_function_schema.py:126]: I added one test where we widen the types of start and end of aten::slice which actually caused production issue before :)

File: test/test_function_schema.py (hunk 12 above)
Comment (tugsbayasgalan) [reply test/test_function_schema.py:136]: good test case. I will add it.

File: test/test_function_schema.py (hunk 11 above)
Comment (tugsbayasgalan) [reply test/test_function_schema.py:120]: will add a test for it

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

File: aten/src/ATen/core/function_schema.h (hunk 6 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/core/function_schema.h:211]: done

File: test/test_function_schema.py (hunk 9 above)
Comment (tugsbayasgalan) [reply test/test_function_schema.py:116]: done

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

**Review by mruberry — APPROVED** [review]
LGTM!

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

**Pull request closed without merging**